spi_nor: devicetree, driver, and sample cleanup

pabigot opened the pull request:
This PR addresses a variety of problems with the spi-nor infrastructure and samples.  It supersedes #17652, an earlier attempt that didn't address all the problems.

The devicetree binding and driver were designed to allow several sizes to be configurable.  The page, sector, erase, and write sizes are all fixed for any device implementing the JEDEC command set; they cannot and should not be configurable.  Driver, devicetree, and bindings have been updated correspondingly.

A couple existing boards have been extended with the pieces necessary to actually use their on-board flash devices in Zephyr.

Samples have been updated to confirm that they work, including the necessary board overrides.  This includes the standard spi_flash sample, and the use of a SPI NOR flash partition as a FCB-based settings store.

Finally, the JEDEC id DTS type has been changed to be a bytestring `[c2 28 17]` rather than an array `<0xc2 0x28 0x17>`.

I can see only two targets--`intel_s1000_crb` and `arduino_101`--that I believe have probably been using the SPI NOR driver.  These should still work correctly, but I don't have the hardware to test them.

pabigot committed:
samples/drivers/spi_flash: fix device naming

There are multiple flash implementations.  This sample was originally
written for the W25QXXDV driver, but has since been used for
jedec,spi-nor.  The sample should work with multiple drivers.

Add device name sources for several drivers, falling back to spi-nor
which was recently used as a test filter.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot committed:
boards/particle_*: correct spi flash description

Particle released documentation with a pre-release flash chip.  Correct
the name to the actual as-sold device, and add the corresponding size
property as well as the has-be32k property.  Also add an alias so we
can set partitions externally.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot committed:
samples/drivers/spi_flash: add board overrides

Particle mesh feather devices, and Nordic PCA10056, can both run
this example with the on-board JEDEC SPI NOR device if the driver
is enabled.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot committed:
drivers/spi_nor: remove configurability of page/sector/block sizes

The JEDEC API defines the hardware page, sector, and block sizes.
Deprecate the Kconfig settings, remove the `erase-size-block` property,
and add `has-be32k` to indicate that 32K-byte erase is supported.
Rework the driver to use the constants instead of configured values.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot committed:
drivers/spi_nor: remove write-block-size devicetree property

Devices using this driver do not require any special alignment for
writes.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot committed:
subsys/storage/flash_map: detect missing flash device

If the flash device is not configured, return an error rather than
dereferencing a null device pointer.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot committed:
tests/subsys/settings/functional: ensure settings area is cleared

The test assumes that the settings area is empty.  This happens
naturally when the area is in the SOC flash and the chip flash is
entirely erased when the test is programmed.  The test will fail if the
programmed board is reset and the test run again.

The test cannot pass by reprogramming when the storage area is in an
external flash device.

Make things repeatable for FCB by erasing the storage partition before
running the test.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot committed:
tests/subsys/settings/functional: test external flash

Use the external flash on the Particle Xenon to confirm basic
functionality on a spi-nor device.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot committed:
drivers/spi_nor: allow application control of flash layout page size

The driver historically used the erase block size (64 KiBy) as the page
size.  There are other viable "sector" sizes, and for some applications
this one may be too large.  Allow the application to specify the desired
flash page size.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot committed:
boards/arm/nrf52840_pca10056: add alias for external flash

Provide an alias so we can assign partitions to this device through
application-specific overlays.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot commented:
One shippable failure appears to require the fix for #17661; the other is just weird.

pabigot committed:
dts/spi-nor: use bytestring for JEDEC ID

This was always intended to be a bytestring rather than an array, but
full support was missing.  Since that has been addressed switch it to
the preferred format.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot commented:
Added the switch of jedec-id to bytestring encoding.  Conversion was automated to reduce typos.
```
git grep -l jedec-id  \
  | xargs sed -i -r \
    -e '/jedec-id/y@<>@[]@' \
    -e '/jedec-id/s@0x@@g'
```

carlescufi commented on tests/subsys/settings/functional/particle_xenon.overlay:
```diff
@@ -0,0 +1,28 @@
+		#address-cells = <1>;
+		#size-cells = <1>;
+
+		littlefs_partition: partition@0 {
```
We don't natively support littlefs in Zephyr. Perhaps it would be better to either make this neutral or else add support (via west) for this filesystem? We know of serious shortcomings with NFFS that need to be addressed, so adding support for littlefs would be great.

pabigot replied:
I'm already a fair way along to littlefs, based on @jimparis's work, but in this case the Particle devices come with a littlefs v1 partition in that location.  So it's correct as is, whether or not we add it to Zephyr.

pabigot committed:
drivers/spi_nor: fix writes across page boundaries

The previous code limited the length of a write to the size of a page,
but did not check whether the starting position was far enough into the
page that the write would still cross a page boundary.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot committed:
drivers/spi_nor: enable logging API

Just register it.  For now it's only used in one place where it was
needed to identify a problem.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

gon1332 commented:
Hi everyone, I strongly believe that at least fde373313a24eaae25562c816babc6d0e2e0c23a (`drivers/spi_nor: fix writes across page boundaries`) should be backported in v1.14.

pabigot commented:
@gon1332 I don't believe backporting the entire PR is feasible as it includes devicetree content changes, but a separate fix for the incorrect behavior seems reasonable.  Could you open a new issue for the backport so we can track it?

gon1332 commented:
> @gon1332 I don't believe backporting the entire PR is feasible as it includes devicetree content changes, but a separate fix for the incorrect behavior seems reasonable. Could you open a new issue for the backport so we can track it?

Totally agree.
